### DISCUSSION TITLE
selecting rows on large ag-grid

### DISCUSSION THREAD

**Discussion by rroyer-xyz** [elimintz/justpy#430] [Ideas]
I have a large grid (over 300k rows) where users can select multiple rows. I'm currently tracking the selections by doing something similar as below. The issue is that there's a significant delay between clicks on the row. You should see this delay when running the example below.

Any suggestions on what to do? Couple things I could think of are...

- Somehow prevent AgGrid and all the rowData from being included in the msg.target when a row is selected
- Use the getSelectedRows() API call to return all selected rows once the user is done instead of tracking each click like below

Thanks

```
import justpy as jp
import pandas as pd

# dummy data
data = {'col1':[], 'col2':[], 'col3':[], 'col4':[], 'col5':[], 'col6':[]} 
for i in range(0,300000):
    data['col1'].append(i)
    data['col2'].append(i)
    data['col3'].append(i)
    data['col4'].append(i)
    data['col5'].append(i)                
    data['col6'].append(i)        

df = pd.DataFrame(data) 

def row_selected(self, msg):
    if msg.selected:
        self.row_data_div.text = msg.rowIndex
        self.row_selected = msg.rowIndex
    elif self.row_selected == msg.rowIndex:
        self.row_data_div.text = ''

def grid_test():

    wp = jp.WebPage()
    row_data_div = jp.Div(a=wp)
    grid = jp.AgGrid(a=wp)
    grid.load_pandas_frame(df) 
    grid.row_data_div = row_data_div
    grid.on('rowSelected', row_selected)   
    grid.options.columnDefs[0].checkboxSelection = True
 
    return wp

jp.justpy(grid_test)
```

**Comment by elimintz** [comment]
Your suggestions make sense. I would add one more suggestion: Perhaps segment the grid and have buttons that load a range of rows. For example, only have 1000 rows in the grid at a time. This is the solution I usually adopt when dealing with large grids.

**Comment by rroyer-xyz** [comment]
Thanks, I think preventing AgGrid (or the rowData) from being included in msg.target might be my best option for now. Could you point me in the direction of where I should start poking around to make this update?

As for using getSelectedRows, I need that data back in python (something like below), and I know that's not possible at the moment in JustPy
`MYDATA = await msg.page.run_javascript(f"""cached_grid_def['g' + {self.grid.id}].api.getSelectedRows()"""
`
Also, I'll check with the users to see if there's a way in this use case to load segments at a time as you mentioned.

**Comment by elimintz** [comment]
rowData is not included in msg.target. You see it when it is printed because of the __repr__ method of AgGrid. The delay is because the browser is having problems handling such a large grid.

**Comment by WolfgangFahl** [comment]
300 k rows? You might want to avoid such a big grid. Otherwise it's a feature request and goes to ideas. Please add more comments if your are still seriously interested and provide pull request or participate in the project by getting invited.
